modules/bam2fastq/1.1/bam2fastq_grouped.smk

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env snakemake
+
+
+##### ATTRIBUTION #####
+
+
+# Original Author:  Lauren Chong
+# Module Author:    Helena Winata
+# Contributors:     N/A
+
+
+##### SETUP #####
+
+
+# Import package with useful functions for developing analysis modules
+import oncopipe as op
+
+# Setup module and store module-specific configuration in `CFG`
+# `CFG` is a shortcut to `config["lcr-modules"]["bam2fastq"]`
+CFG = op.setup_module(
+    name = "bam2fastq",
+    version = "1.1",
+    subdirectories = ["inputs", "fastq", "outputs"],
+)
+
+# Define rules to be run locally when using a compute cluster
+localrules:
+    _bam2fastq_input_bam,
+    _bam2fastq_output,
+    _bam2fastq_all,
+
+
+##### RULES #####
+
+# Functions to obtain the genome_build per sample from the sample table
+def get_genome_build(wildcards):
+    tbl = config["lcr-modules"]["bam2fastq"]["samples"]
+    genome_build = tbl[(tbl.sample_id == wildcards.sample_id) & (tbl.seq_type == wildcards.seq_type)]["genome_build"]
+    return reference_files(expand("genomes/{genome_build}/genome_fasta/genome.fa", genome_build = genome_build))
+
+def get_bams(wildcards,build = False):
+    CFG = config["lcr-modules"]["bam2fastq"]
+    tbl = CFG["samples"]
+    return(expand(
+        CFG["dirs"]["inputs"] + "{{seq_type}}--{genome_build}/{{sample_id}}.bam", 
+        genome_build = tbl[(tbl.sample_id == wildcards.sample_id) & (tbl.seq_type == wildcards.seq_type)]["genome_build"]
+    ))
+
+
+rule _bam2fastq_input_bam:
+    input:
+        bam_path = CFG['inputs']['sample_bam']
+    output:
+        bam = CFG["dirs"]["inputs"] + "{seq_type}--{genome_build}/{sample_id}.bam"
+    group: 
+        CFG["group"]["bam2fastq"]
+    run:
+        op.relative_symlink(input, output.bam)
+
+
+# Conditional rules depending on whether or not fastq outputs will be temporary
+if CFG["temp_outputs"] == True:
+    rule _bam2fastq_run:
+        input:
+            bam = get_bams,
+            genome = get_genome_build
+        output:
+            fastq_1 = temp(CFG["dirs"]["fastq"] + "{seq_type}/{sample_id}.read1.fastq.gz"),
+            fastq_2 = temp(CFG["dirs"]["fastq"] + "{seq_type}/{sample_id}.read2.fastq.gz")
+        log:
+            stdout = CFG["logs"]["fastq"] + "{seq_type}/{sample_id}/bam2fastq.stdout.log",
+            stderr = CFG["logs"]["fastq"] + "{seq_type}/{sample_id}/bam2fastq.stderr.log"
+        params:
+            opts = CFG["options"]["bam2fastq"]
+        conda:
+            CFG["conda_envs"]["picard"]
+        threads:
+            CFG["threads"]["bam2fastq"]
+        resources:
+            **CFG["resources"]["bam2fastq"]
+        group: 
+            CFG["group"]["bam2fastq"]
```

Comment:
Does the grouping also needs to be added if fastq files are not marked as temporary?

Reply:
I can't really envision a scenario when it would be needed, but this still gives users the flexibility.